### PR TITLE
Introduce RequestTransformer chaining

### PR DESCRIPTION
**Pull request by jrodbx**

### CODE REVIEW

File: picasso/src/main/java/com/squareup/picasso/Picasso.java [hunk 1]
```diff
@@ -471,14 +472,23 @@ public void shutdown() {
...
   Request transformRequest(Request request) {
-    Request transformed = requestTransformer.transformRequest(request);
-    if (transformed == null) {
-      throw new IllegalStateException("Request transformer "
-          + requestTransformer.getClass().getCanonicalName()
-          + " returned null for "
-          + request);
+    if (requestTransformers == null) {
+      return request;
     }
```
Comment (jrodbx) [inline picasso/src/main/java/com/squareup/picasso/Picasso.java]: I wrestled between this and an empty ArrayList().  Effective Java says the latter, but Picasso has a history of conservative memory usage.  Side effect: test setup is cleaner.  Thoughts?

Comment (NightlyNexus) [reply picasso/src/main/java/com/squareup/picasso/Picasso.java]: Collections.emptyList() has no memory effect.

Comment (NightlyNexus) [reply picasso/src/main/java/com/squareup/picasso/Picasso.java]: oh, i see. in the builder. yeah, null seems fine.

File: picasso/src/main/java/com/squareup/picasso/Picasso.java [hunk 2]
```diff
@@ -768,20 +778,18 @@ public Builder listener(@NonNull Listener listener) {
...
+    /** Add a transformer that observes and potentially modify all incoming requests. */
+    public Builder addRequestTransformer(RequestTransformer transformer) {
       if (transformer == null) {
         throw new IllegalArgumentException("Transformer must not be null.");
       }
-      if (this.transformer != null) {
+      if (requestTransformers == null) {
+        requestTransformers = new ArrayList<>();
+      }
+      if (requestTransformers.contains(transformer)) {
```
Comment (NightlyNexus) [inline picasso/src/main/java/com/squareup/picasso/Picasso.java]: do we care about this case, anyway?

File: picasso/src/main/java/com/squareup/picasso/Picasso.java [hunk 3]
```diff
@@ -149,7 +149,7 @@
   };
 
   private final Listener listener;
-  private final RequestTransformer requestTransformer;
+  private final List<RequestTransformer> requestTransformers;
```
Comment (NightlyNexus) [inline picasso/src/main/java/com/squareup/picasso/Picasso.java]: `@Nullable`

File: picasso/src/main/java/com/squareup/picasso/Picasso.java [hunk 4]
```diff
@@ -168,13 +168,14 @@
   boolean shutdown;
 
   Picasso(Context context, Dispatcher dispatcher, Cache cache, Listener listener,
-      RequestTransformer requestTransformer, List<RequestHandler> extraRequestHandlers, Stats stats,
-      Bitmap.Config defaultBitmapConfig, boolean indicatorsEnabled, boolean loggingEnabled) {
+      List<RequestTransformer> requestTransformers, List<RequestHandler> extraRequestHandlers,
```
Comment (NightlyNexus) [inline picasso/src/main/java/com/squareup/picasso/Picasso.java]: `@Nullable` on the list param.

File: picasso/src/main/java/com/squareup/picasso/Picasso.java [hunk 5]
```diff
@@ -168,13 +168,14 @@
...
+      Stats stats, Bitmap.Config defaultBitmapConfig, boolean indicatorsEnabled,
+      boolean loggingEnabled) {
     this.context = context;
     this.dispatcher = dispatcher;
     this.cache = cache;
     this.listener = listener;
-    this.requestTransformer = requestTransformer;
+    this.requestTransformers = Collections.unmodifiableList(new ArrayList<>(requestTransformers));
```
Comment (NightlyNexus) [inline picasso/src/main/java/com/squareup/picasso/Picasso.java]: refactoring bug?
`this.requestTransformers = requestTransformers != null ? Collections.unmodifiableList(new ArrayList<>(requestTransformers)) : null;`

File: picasso/src/main/java/com/squareup/picasso/Picasso.java (hunk 2 above)
Comment (jrodbx) [reply picasso/src/main/java/com/squareup/picasso/Picasso.java]: depends on the transformer impl?

File: picasso/src/main/java/com/squareup/picasso/Picasso.java (hunk 5 above)
Comment (jrodbx) [reply picasso/src/main/java/com/squareup/picasso/Picasso.java]: yep, good catch.

File: picasso/src/main/java/com/squareup/picasso/Picasso.java (hunk 2 above)
Comment (NightlyNexus) [reply picasso/src/main/java/com/squareup/picasso/Picasso.java]: i mean how likely is it that somebody accidentally adds the same RequestTransformer twice?

Comment (jrodbx) [reply picasso/src/main/java/com/squareup/picasso/Picasso.java]: bugs happen :)

File: picasso/src/main/java/com/squareup/picasso/Picasso.java (hunk 1 above)
Comment (NightlyNexus) [reply picasso/src/main/java/com/squareup/picasso/Picasso.java]: sorry, just realized this is Picasso.Builder. was thinking about the Request.Builder.
let's do a non-null list. you're just creating a shared Picasso instance.

Comment (NightlyNexus) [reply picasso/src/main/java/com/squareup/picasso/Picasso.java]: sorry about that.

**Comment by NightlyNexus** [comment]
i think somebody said somewhere, but remind me again what the motivation is.

could a user write a single RequestTransformer, moving the loop to user code instead?
or, is this in preparation for more functionality in the chain?

File: picasso/src/main/java/com/squareup/picasso/Picasso.java [hunk 6]
```diff
@@ -471,14 +473,23 @@ public void shutdown() {
...
-      throw new IllegalStateException("Request transformer "
-          + requestTransformer.getClass().getCanonicalName()
-          + " returned null for "
-          + request);
+    if (requestTransformers == null) {
+      return request;
     }
-    return transformed;
+
+    int numTransformers = requestTransformers.size();
```
Comment (JakeWharton) [inline picasso/src/main/java/com/squareup/picasso/Picasso.java]: I typically put this into the loop
```
for (int i = 0, size = requestTransformers.size(); i < size; i++) {
```

**Comment by jrodbx** [comment]
@NightlyNexus 
yep, Jake mentioned it here: https://github.com/square/picasso/issues/1800#issuecomment-371693216.  The idea is to mimic OkHttp's interceptor model and allow each transformer to potentially modify requests as it propagates down the chain.

File: picasso/src/main/java/com/squareup/picasso/Picasso.java (hunk 6 above)
Comment (jrodbx) [reply picasso/src/main/java/com/squareup/picasso/Picasso.java]: updated.

File: picasso/src/main/java/com/squareup/picasso/Picasso.java (hunk 1 above)
Comment (jrodbx) [reply picasso/src/main/java/com/squareup/picasso/Picasso.java]: updated.

File: picasso/src/main/java/com/squareup/picasso3/Picasso.java [hunk 7]
```diff
@@ -678,7 +674,7 @@ void shutdown() {
     private ExecutorService service;
     private Cache cache;
     private Listener listener;
-    private RequestTransformer transformer;
+    private List<RequestTransformer> requestTransformers = new ArrayList<>();
```
Comment (NightlyNexus) [inline picasso/src/main/java/com/squareup/picasso3/Picasso.java]: nit: can be `final`

Comment (jrodbx) [reply picasso/src/main/java/com/squareup/picasso3/Picasso.java]: updated

**Pull request merged**